### PR TITLE
feat: put common schema types into base package

### PR DESCRIPTION
**Pull request by kalabukdima**

### CODE REVIEW

File: packages/mangrove/src/eth.ts [hunk 1]
```diff
@@ -1,12 +1,7 @@
...
 export type UInt = string;
-export type Hash = string;
-export type BlockNumber = number;
 export type Int = string;
 
-export type TransactionRef = {
-  blockNumber: BlockNumber;
-  blockHash: Hash;
-  txHash: Hash;
-  from: Address;
```
Comment (kalabukdima) [inline packages/mangrove/src/eth.ts:11]: `from` is renamed to `sender` here. It seems to be only used in [mangrove-offchain-server](https://github.com/proxima-one/mangrove-offchain-server/blob/706c57817f730b5b89688f8ebe74abb18c61c7ff/src/state/mangrove/handler.ts#L58) repo. I can simply follow with a fixing commit there or I can add optional `from` field here to preserve the compatibility.

File: packages/base/src/blockchainSchema.ts [hunk 2]
```diff
@@ -0,0 +1,17 @@
+import { JsonObject, JsonValue } from "./jsonSchema";
+
+export type Hash = string;
+export type Amount = string;
+export type Address = Hash;
+export type UserId = Hash;
```
Comment (kalabukdima) [inline packages/base/src/blockchainSchema.ts:6]: Or should we just use `Address` instead?

File: packages/mangrove/src/eth.ts [hunk 3]
```diff
@@ -1,12 +1,7 @@
...
-};
+export type TransactionRef = TxRef;
```
Comment (kalabukdima) [inline packages/mangrove/src/eth.ts:7]: Preserved here for compatibility. However, in the new code we should use `TxRef` directly.

File: packages/base/src/blockchainSchema.ts (hunk 2 above)
Comment (ivanbenko) [reply packages/base/src/blockchainSchema.ts:6]: `UserId` is not hash for all blockchains, so just leaving it as `string` alias makes more sense. Same for `Address`.

ps:
Actually it looks like `UserId` doesn't fit very well to generic blockchain concept. Even for transfer where we use it for `from` -> `to` it's not always a user, but f.e. smart contract. So probably just leaving `Address` or renaming `User` to `Account` would make more sense

File: packages/base/src/blockchainSchema.ts [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+
+
+interface TxRefBase<TBlockNumber extends JsonValue> extends JsonObject {
+  blockNumber?: TBlockNumber;
+  blockHash?: Hash;
+  txHash?: Hash;
+  sender?: UserId;
+}
```
Comment (ivanbenko) [inline packages/base/src/blockchainSchema.ts:14]: Having the same structure for `TxRef` is fine, but I don't like that all properties are optional what forces consumer to check whether they all set. I suggest rather:
```typescript

interface TxRefBase<TBlockNumber extends JsonValue> extends JsonObject {
  chain: string;
  blockNumber: TBlockNumber;
  blockHash: Hash;
  txHash: Hash;
  sender: UserId;
}

export type TxRef = TxRefBase<number>;
export type TxRefStr = TxRefBase<string>;
```

And every stream/protocol can decide how it's `TxRef` should look like, so every stream has it's strongly typed TxRef:

```typescript
type UniswapTxRef = Pick<TxRef, "blockNumber" | "blockHash" | "txHash">;
type NearTxRef = Omit<TxRef, "blockNumber" | "sender">;
```

**Review by ivanbenko — CHANGES_REQUESTED** [review]

**Pull request merged**